runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java

```diff
@@ -0,0 +1,382 @@
+      RunnerApi.Pipeline pipeline, List<String> urnsToOverride, ExternalTranslationOptions options)
+      throws Exception {
+    List<String> transformsToOverride =
+        pipeline.getComponents().getTransformsMap().entrySet().stream()
+            .filter(
+                entry -> {
+                  String urn = entry.getValue().getSpec().getUrn();
+                  if (urn != null && urnsToOverride.contains(urn)) {
+                    return true;
+                  }
+                  return false;
+                })
+            .map(
+                entry -> {
+                  return entry.getKey();
+                })
+            .collect(Collectors.toList());
+
+    String serviceAddress;
+    TransformServiceLauncher service = null;
+    try {
+      if (options.getTransformServiceAddress() != null) {
+        serviceAddress = options.getTransformServiceAddress();
+      } else if (options.getTransformServiceBeamVersion() != null) {
+        String projectName = UUID.randomUUID().toString();
+        int port = findAvailablePort();
+        service = TransformServiceLauncher.forProject(projectName, port);
+        service.setBeamVersion(options.getTransformServiceBeamVersion());
+
+        // Starting the transform service.
+        service.start();
+        service.waitTillUp(40);
+        serviceAddress = "localhost:" + Integer.toString(port);
+        System.out.println("Done waiting ...");
+      } else {
+        throw new IllegalArgumentException(
+            "Either option TransformServiceAddress or option TransformServiceBeamVersion should be "
+                + "provided to override a transform using the transform service");
+      }
+
+      Endpoints.ApiServiceDescriptor expansionServiceEndpoint =
+          Endpoints.ApiServiceDescriptor.newBuilder().setUrl(serviceAddress).build();
+
+      for (String transformId : transformsToOverride) {
+        pipeline =
+            updateTransformViaTransformService(pipeline, transformId, expansionServiceEndpoint);
+      }
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+
+    if (service != null) {
+      try {
+        service.shutdown();
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    return pipeline;
+  }
+
+  private <
+          InputT extends PInput,
+          OutputT extends POutput,
+          TransformT extends org.apache.beam.sdk.transforms.PTransform<InputT, OutputT>>
+      RunnerApi.Pipeline updateTransformViaTransformService(
+          RunnerApi.Pipeline runnerAPIpipeline,
+          String transformId,
+          Endpoints.ApiServiceDescriptor transformServiceEndpoint)
+          throws Exception {
+    PTransform transformToUpgrade =
+        runnerAPIpipeline.getComponents().getTransformsMap().get(transformId);
+    if (transformToUpgrade == null) {
+      throw new Exception("Could not find a transform with the ID " + transformId);
```

Comment:
Seems like a good candidate for `Preconditions.checkNotNull()`. Or throw something like `IllegalArgumentException`.